sources/api/early-boot-config/src/provider/vmware.rs

```diff
@@ -0,0 +1,397 @@
+                    path: &user_data_file,
+                })?
+            }
+        };
+
+        if user_data_str.is_empty() {
+            return Ok(None);
+        }
+
+        // User data could be 700MB compressed!  Eek!  :)
+        if user_data_str.len() <= 2048 {
+            trace!("Received user data: {}", user_data_str);
+        } else {
+            trace!(
+                "Received long user data, starts with: {}",
+                // (this isn't perfect because chars aren't grapheme clusters, but will error
+                // toward printing the whole input, which is fine)
+                String::from_iter(user_data_str.chars().take(2048))
+            );
+        }
+
+        let json = SettingsJson::from_toml_str(&user_data_str, "user data from CD-ROM").context(
+            error::SettingsToJSON {
+                from: user_data_file.display().to_string(),
+            },
+        )?;
+
+        Ok(Some(json))
+    }
+
+    /// Read and base64 decode user data contained in an OVF file
+    // In VMWare, user data is supplied to the host via an XML file.  Within
+    // the XML file, there is a `PropertySection` that contains `Property` elements
+    // with attributes.  User data is base64 encoded inside a `Property` element with
+    // the attribute "user-data".
+    // <Property key="user-data" value="1234abcd"/>
+    fn ovf_user_data<P: AsRef<Path>>(path: P) -> Result<String> {
+        let path = path.as_ref();
+        let file = File::open(path).context(error::InputFileRead { path })?;
+        let reader = OptionalCompressionReader::new(BufReader::new(file));
+
+        // Deserialize the OVF file, dropping everything we don't care about
+        let ovf: Environment =
+            serde_xml_rs::from_reader(reader).context(error::XmlDeserialize { path })?;
+
+        // We have seen the keys in the `Property` section be "namespaced" like "oe:key" or
+        // "of:key".  Since we aren't trying to validate the schema beyond the presence of the
+        // elements we care about, we can ignore the namespacing.  An example of this type of
+        // namespacing can be found in the unit test sample data. `serde_xml_rs` effectively
+        // ignores these namespaces and returns "key" / "value":
+        // https://github.com/Rreverser/serde-xml-rs/issues/64#issuecomment=540448434
+        let mut base64_str = String::new();
+        let user_data_key = "user-data";
+        for property in ovf.property_section.properties {
+            if property.key == user_data_key {
+                base64_str = property.value;
+                break;
+            }
+        }
+
+        // Base64 decode the &str
+        let decoded_bytes = base64::decode(&base64_str).context(error::Base64Decode {
+            base64_string: base64_str.to_string(),
+        })?;
+
+        // Decompress the data if it's compressed
+        let decoded = expand_slice_maybe(&decoded_bytes).context(error::Decompression {
```

Comment:
Here's another spot we could chain the OptionalCompressionReader and DecoderReader.

One note though - the error is no longer strictly related to decompression - it could fail base64 too - so I think we'd want a new variant that's a little more generic, maybe UserDataInput or something?